<a href="https://cognitiveclass.ai"><img src = "https://ibm.box.com/shared/static/9gegpsmnsoo25ikkbl4qzlvlyjbgxs5x.png" width = 400> </a>

<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto</font></h1>


Before we get the data and start exploring it, let's download all the dependencies that we will need.


# Part 1

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


<a id='item1'></a>


## 1. Download and Explore Dataset


In [70]:
#downloaded the files and scrapped the data out wikipedia page 
from IPython.display import display_html
source = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
soup=BeautifulSoup(source,'lxml')
print(soup.title)
tab = str(soup.table)
display_html(tab, raw=True)

<title>List of postal codes of Canada: M - Wikipedia</title>


Postal Code
 
 Borough
 
 Neighbourhood
 
 
 M1A
 
 Not assigned
 
 Not assigned
 
 
 M2A
 
 Not assigned
 
 Not assigned
 
 
 M3A
 
 North York
 
 Parkwoods
 
 
 M4A
 
 North York
 
 Victoria Village
 
 
 M5A
 
 Downtown Toronto
 
 Regent Park, Harbourfront
 
 
 M6A
 
 North York
 
 Lawrence Manor, Lawrence Heights
 
 
 M7A
 
 Downtown Toronto
 
 Queen's Park, Ontario Provincial Government
 
 
 M8A
 
 Not assigned
 
 Not assigned
 
 
 M9A
 
 Etobicoke
 
 Islington Avenue, Humber Valley Village
 
 
 M1B
 
 Scarborough
 
 Malvern, Rouge
 
 
 M2B
 
 Not assigned
 
 Not assigned
 
 
 M3B
 
 North York
 
 Don Mills
 
 
 M4B
 
 East York
 
 Parkview Hill, Woodbine Gardens
 
 
 M5B
 
 Downtown Toronto
 
 Garden District, Ryerson
 
 
 M6B
 
 North York
 
 Glencairn
 
 
 M7B
 
 Not assigned
 
 Not assigned
 
 
 M8B
 
 Not assigned
 
 Not assigned
 
 
 M9B
 
 Etobicoke
 
 West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
 
 
 M1C
 
 Scarborough
 
 Rouge Hill, Port Union, Highland Creek
 
 
 M2C
 
 Not assigned
 
 Not assigned
 
 
 M3C
 
 North York
 
 Don Mills
 
 
 M4C
 
 East York
 
 Woodbine Heights
 
 
 M5C
 
 Downtown Toronto
 
 St. James Town
 
 
 M6C
 
 York
 
 Humewood-Cedarvale
 
 
 M7C
 
 Not assigned
 
 Not assigned
 
 
 M8C
 
 Not assigned
 
 Not assigned
 
 
 M9C
 
 Etobicoke
 
 Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
 
 
 M1E
 
 Scarborough
 
 Guildwood, Morningside, West Hill
 
 
 M2E
 
 Not assigned
 
 Not assigned
 
 
 M3E
 
 Not assigned
 
 Not assigned
 
 
 M4E
 
 East Toronto
 
 The Beaches
 
 
 M5E
 
 Downtown Toronto
 
 Berczy Park
 
 
 M6E
 
 York
 
 Caledonia-Fairbanks
 
 
 M7E
 
 Not assigned
 
 Not assigned
 
 
 M8E
 
 Not assigned
 
 Not assigned
 
 
 M9E
 
 Not assigned
 
 Not assigned
 
 
 M1G
 
 Scarborough
 
 Woburn
 
 
 M2G
 
 Not assigned
 
 Not assigned
 
 
 M3G
 
 Not assigned
 
 Not assigned
 
 
 M4G
 
 East York
 
 Leaside
 
 
 M5G
 
 Downtown Toronto
 
 Central Bay Street
 
 
 M6G
 
 Downtown Toronto
 
 Christie
 
 
 M7G
 
 Not assigned
 
 Not assigned
 
 
 M8G
 
 Not assigned
 
 Not assigned
 
 
 M9G
 
 Not assigned
 
 Not assigned
 
 
 M1H
 
 Scarborough
 
 Cedarbrae
 
 
 M2H
 
 North York
 
 Hillcrest Village
 
 
 M3H
 
 North York
 
 Bathurst Manor, Wilson Heights, Downsview North
 
 
 M4H
 
 East York
 
 Thorncliffe Park
 
 
 M5H
 
 Downtown Toronto
 
 Richmond, Adelaide, King
 
 
 M6H
 
 West Toronto
 
 Dufferin, Dovercourt Village
 
 
 M7H
 
 Not assigned
 
 Not assigned
 
 
 M8H
 
 Not assigned
 
 Not assigned
 
 
 M9H
 
 Not assigned
 
 Not assigned
 
 
 M1J
 
 Scarborough
 
 Scarborough Village
 
 
 M2J
 
 North York
 
 Fairview, Henry Farm, Oriole
 
 
 M3J
 
 North York
 
 Northwood Park, York University
 
 
 M4J
 
 East York
 
 East Toronto, Broadview North (Old East York)
 
 
 M5J
 
 Downtown Toronto
 
 Harbourfront East, Union Station, Toronto Islands
 
 
 M6J
 
 West Toronto
 
 Little Portugal, Trinity
 
 
 M7J
 
 Not assigned
 
 Not assigned
 
 
 M8J
 
 Not assigned
 
 Not assigned
 
 
 M9J
 
 Not assigned
 
 Not assigned
 
 
 M1K
 
 Scarborough
 
 Kennedy Park, Ionview, East Birchmount Park
 
 
 M2K
 
 North York
 
 Bayview Village
 
 
 M3K
 
 North York
 
 Downsview
 
 
 M4K
 
 East Toronto
 
 The Danforth West, Riverdale
 
 
 M5K
 
 Downtown Toronto
 
 Toronto Dominion Centre, Design Exchange
 
 
 M6K
 
 West Toronto
 
 Brockton, Parkdale Village, Exhibition Place
 
 
 M7K
 
 Not assigned
 
 Not assigned
 
 
 M8K
 
 Not assigned
 
 Not assigned
 
 
 M9K
 
 Not assigned
 
 Not assigned
 
 
 M1L
 
 Scarborough
 
 Golden Mile, Clairlea, Oakridge
 
 
 M2L
 
 North York
 
 York Mills, Silver Hills
 
 
 M3L
 
 North York
 
 Downsview
 
 
 M4L
 
 East Toronto
 
 India Bazaar, The Beaches West
 
 
 M5L
 
 Downtown Toronto
 
 Commerce Court, Victoria Hotel
 
 
 M6L
 
 North York
 
 North Park, Maple Leaf Park, Upwood Park
 
 
 M7L
 
 Not assigned
 
 Not assigned
 
 
 M8L
 
 Not assigned
 
 Not assigned
 
 
 M9L
 
 North York
 
 Humber Summit
 
 


Create dataframe from our data 

In [64]:
dfs = pd.read_html(tab)
df=dfs[0]
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [72]:
# dropping cells with a borough that is Not assigned.
df1 = df[df.Borough != 'Not assigned']

# Combining the neighbourhoods that shares the same Postalcode
toronto_df = df1.groupby(['Postal Code','Borough'], sort=False).agg(', '.join)
toronto_df.reset_index(inplace=True)


#Making the neighborhood the same as the borough, a cell has a borough but a Not assigned 
toronto_df['Neighbourhood'] = np.where(df2['Neighbourhood'] == 'Not assigned',df2['Borough'], df2['Neighbourhood'])


In [73]:
toronto_df.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [74]:
# shape method to print the number of rows of your dataframe.
toronto_df.shape

(103, 3)

# Part 2

Using the Geocoder package or the csv file to add 'Latitude' and 'Longitutde' to the toronto_df

In [75]:
lat_lon = pd.read_csv('https://cocl.us/Geospatial_data')
lat_lon.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [76]:
toronto_df_new = pd.merge(toronto_df,lat_lon,on='Postal Code')
toronto_df_new.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

# Part 3 

#### Load and explore the data


Next, let's load the data.


Let's take a quick look at the data.


And make sure that the dataset has all 10 boroughs and 103 neighborhoods.


In [80]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(toronto_df_new['Borough'].unique()),
        toronto_df_new.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


#### Use geopy library to get the latitude and longitude values of Toronto


In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>Toronto Explorer</em>, as shown below.


In [81]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="Toronto Explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a map of Toronto Explorer with neighborhoods superimposed on top.


In [84]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_df_new['Latitude'], toronto_df_new['Longitude'], toronto_df_new['Borough'], toronto_df_new['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [87]:
Downtown_Toronto_data = toronto_df_new[toronto_df_new['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
Downtown_Toronto_data.head()

Postal Code           Borough                                Neighbourhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

Let's get the geographical coordinates of Downtown Toronto.


In [ ]:
address = 'Downtown Toronto, ON'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

As we did with all of Toronto, let's visualizat Downtown Toronto the neighborhoods in it

In [88]:
# create map of Toronto using latitude and longitude values
map_Downtown_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(Downtown_Toronto_data['Latitude'], Downtown_Toronto_data['Longitude'], Downtown_Toronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Downtown_Toronto)  
    
map_Downtown_Toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.


#### Define Foursquare Credentials and Version


In [89]:
CLIENT_ID = 'GSBPKOLYKWIXWODEAEDMKMO0H0DHPRUQ1RLLKXT4OFGS0HJG' 
CLIENT_SECRET = 'PILEGZLMDCJW3RD4CBK1CAWYRWQFVAADYHAMI3QMDHENWRBD'
VERSION = '20180605' 
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: GSBPKOLYKWIXWODEAEDMKMO0H0DHPRUQ1RLLKXT4OFGS0HJG
CLIENT_SECRET:PILEGZLMDCJW3RD4CBK1CAWYRWQFVAADYHAMI3QMDHENWRBD


#### Let's explore the first neighborhood in our dataframe.


Get the neighborhood's name.


In [91]:
Downtown_Toronto_data.loc[0, 'Neighbourhood']

'Regent Park, Harbourfront'

Get the neighborhood's latitude and longitude values.


In [93]:
neighborhood_latitude =Downtown_Toronto_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude =Downtown_Toronto_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = Downtown_Toronto_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


#### Now, let's get the top 100 venues that are in Regent Park within a radius of 500 meters.


First, let's create the GET request URL


In [94]:
# type your answer here

LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url 



'https://api.foursquare.com/v2/venues/explore?&client_id=GSBPKOLYKWIXWODEAEDMKMO0H0DHPRUQ1RLLKXT4OFGS0HJG&client_secret=PILEGZLMDCJW3RD4CBK1CAWYRWQFVAADYHAMI3QMDHENWRBD&v=20180605&ll=43.6542599,-79.3606359&radius=500&limit=100'

Send the GET request and examine the resutls


In [95]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ff9fbd1c8ba5b3c6d14fde3'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 46,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

From the Foursquare lab in the previous module, we know that all the information is in the _items_ key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.


In [96]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a _pandas_ dataframe.


In [97]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

C:\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3      Morning Glory Cafe       Breakfast Spot  43.653947 -79.361149
4     Body Blitz Spa East                  Spa  43.654735 -79.359874

And how many venues were returned by Foursquare?


In [98]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

46 venues were returned by Foursquare.


<a id='item2'></a>


## 2. Explore Neighborhoods in Downtown Toronto


#### Let's create a function to repeat the same process to all the neighborhoods in Toronto


In [99]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called Downtown_Toronto_venues


In [102]:

Downtown_Toronto_venues = getNearbyVenues(names=Downtown_Toronto_data['Neighbourhood'],
                                   latitudes=Downtown_Toronto_data['Latitude'],
                                   longitudes=Downtown_Toronto_data['Longitude']
                                  )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


#### Let's check the size of the resulting dataframe


In [103]:
print(Downtown_Toronto_venues.shape)
Downtown_Toronto_venues.head()

(1226, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3      Morning Glory Cafe       43.653947       -79.361149   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3       Breakfast Spot  
4                  Spa

Let's check how many venues were returned for each neighborhood


In [106]:
Downtown_Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            59   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     61   
Christie                                                               16   
Church and Wellesley                                                   78   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              66   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              46   
Richmond, Adelaide, King                                               96   
Rosedale                                                                4   
St. James Town                                                         79   
St. James Town, Cabbagetown                                            41   
Stn A PO Boxes                                                         98   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         32   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             59   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      61   
Christie                                                                16   
Church and Wellesley                                                    78   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               66   
Queen's Park, Ontario Provincial Government                             33   
Regent Park, Harbourfront                                               46   
Richmond, Adelaide, King                                                96   
Rosedale                                                                 4   
St. James Town                                                          79   
St. James Town, Cabbagetown                                             41   
Stn A PO Boxes                                                          98   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          32   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            59              59   
CN Tower, King and Spadina, Railway Lands, Harb...     17              17   
Central Bay Street                                     61              61   
Christie                                               16              16   
Church and Wellesley                                   78              78   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

#### Let's find out how many unique categories can be curated from all the returned venues


In [107]:
print('There are {} uniques categories.'.format(len(Downtown_Toronto_venues['Venue Category'].unique())))

There are 207 uniques categories.


<a id='item3'></a>


## 3. Analyze Each Neighborhood


In [109]:
# one hot encoding
Downtown_Toronto_onehot = pd.get_dummies(Downtown_Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Downtown_Toronto_onehot['Neighborhood'] = Downtown_Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Downtown_Toronto_onehot.columns[-1]] + list(Downtown_Toronto_onehot.columns[:-1])
Downtown_Toronto_onehot = Downtown_Toronto_onehot[fixed_columns]

Downtown_Toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0            0        0                   0             0               0   
1            0        0                   0             0               0   
2            0        0                   0             0               0   
3            0        0                   0             0               0   
4            0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                   0          0           0           0       1     0    0   
1                   0          0           0           0       0     0    0   
2                   0          0           0           0       0     0    0   
3                   0          0           0           0       0     0    0   
4                   0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0                   0       0              0          0         0   
1           0                   0       0              0          0         0   
2           0                   0       0              0          0         0   
3           0                   0       0              0          0         0   
4           0                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               1        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Butcher  Café  Camera Store  Candy Store  \
0             0              0        0     0             0            0   
1             0              0        0     0             0            0   
2             0              0        0     0             0            0   
3             0              0        0     0             0            0   
4             0              0        0     0             0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0               0   
2                     0      

And let's examine the new dataframe size.


In [110]:
Downtown_Toronto_onehot.shape

(1226, 207)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [111]:
Downtown_Toronto_grouped = Downtown_Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Downtown_Toronto_grouped

Neighborhood  Yoga Studio   Airport  \
0                                         Berczy Park     0.000000  0.000000   
1   CN Tower, King and Spadina, Railway Lands, Har...     0.000000  0.058824   
2                                  Central Bay Street     0.016393  0.000000   
3                                            Christie     0.000000  0.000000   
4                                Church and Wellesley     0.025641  0.000000   
5                      Commerce Court, Victoria Hotel     0.000000  0.000000   
6              First Canadian Place, Underground city     0.000000  0.000000   
7                            Garden District, Ryerson     0.000000  0.000000   
8   Harbourfront East, Union Station, Toronto Islands     0.000000  0.000000   
9           Kensington Market, Chinatown, Grange Park     0.000000  0.000000   
10        Queen's Park, Ontario Provincial Government     0.030303  0.000000   
11                          Regent Park, Harbourfront     0.021739  0.000000   
12                           Richmond, Adelaide, King     0.000000  0.000000   
13                                           Rosedale     0.000000  0.000000   
14                                     St. James Town     0.000000  0.000000   
15                        St. James Town, Cabbagetown     0.024390  0.000000   
16                                     Stn A PO Boxes     0.010204  0.000000   
17           Toronto Dominion Centre, Design Exchange     0.000000  0.000000   
18                     University of Toronto, Harbord     0.031250  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.058824      0.058824        0.117647         0.176471   
2             0.000000      0.000000        0.000000         0.000000   
3             0.000000      0.000000        0.000000         0.000000   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000         0.000000   
9             0.000000      0.000000        0.000000         0.000000   
10            0.000000      0.000000        0.000000         0.000000   
11            0.000000      0.000000        0.000000         0.000000   
12            0.000000      0.000000        0.000000         0.000000   
13            0.000000      0.000000        0.000000         0.000000   
14            0.000000      0.000000        0.000000         0.000000   
15            0.000000      0.000000        0.000000         0.000000   
16            0.000000      0.000000        0.000000         0.000000   
17            0.000000      0.000000        0.000000         0.000000   
18            0.000000      0.000000        0.000000         0.000000   

    Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0           0.000000             0.000000      0.000000      0.00   
1           0.117647             0.000000      0.000000      0.00   
2           0.000000             0.000000      0.000000      0.00   
3           0.000000             0.000000      0.000000      0.00   
4           0.000000             0.012821      0.000000      0.00   
5           0.000000             0.040000      0.000000      0.00   
6           0.000000             0.030000      0.000000      0.00   
7           0.000000             0.000000      0.000000      0.00   
8           0.000000             0.000000      0.000000      0.05   
9           0.000000             0.000000      0.000000      0.00   
10          0.000000             0.000000      0.000000      0.00   
11          0.000000             0.000000      0.021739      0.00   
12          0.000000             0.020833      0.000000      0.00   
13          0

#### Let's confirm the new size


In [113]:
Downtown_Toronto_grouped.shape

(19, 207)

There are 19 neigbourhoods with 207 categories 

#### Let's print each neighborhood along with the top 5 most common venues


In [114]:
num_top_venues = 5

for hood in Downtown_Toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Downtown_Toronto_grouped[Downtown_Toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
          venue  freq
0   Coffee Shop  0.08
1  Cocktail Bar  0.05
2    Restaurant  0.03
3      Beer Bar  0.03
4        Bakery  0.03


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0   Airport Service  0.18
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3   Harbor / Marina  0.06
4          Boutique  0.06


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.20
1   Italian Restaurant  0.05
2                 Café  0.05
3       Sandwich Place  0.05
4  Japanese Restaurant  0.03


----Christie----
                venue  freq
0       Grocery Store  0.25
1                Café  0.19
2                Park  0.12
3         Candy Store  0.06
4  Italian Restaurant  0.06


----Church and Wellesley----
                  venue  freq
0           Coffee Shop  0.08
1   Japanese Restaurant  0.06
2      Sushi Restaurant  0.06
3               Gay Bar  0.04
4  F

#### Let's put that into a _pandas_ dataframe


First, let's write a function to sort the venues in descending order.


In [115]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.


In [118]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Downtown_Toronto_grouped['Neighborhood']

for ind in np.arange(Downtown_Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Downtown_Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar           Cheese Shop            Restaurant   
1        Airport Lounge      Airport Terminal                 Plane   
2        Sandwich Place    Italian Restaurant                  Café   
3                  Café                  Park           Candy Store   
4   Japanese Restaurant      Sushi Restaurant               Gay Bar   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Farmers Market    Seafood Restaurant              Beer Bar   
1       Harbor / Marina      Sculpture Garden         Boat or Ferry   
2       Bubble Tea Shop          Burger Joint           Salad Place   
3            Restaurant    Italian Restaurant            Baby Store   
4            Restaurant  Fast Food Restaurant           Yoga Studio   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0                Bakery        Breakfast Spot                  Irish Pub  
1   Rental Car Location           Coffee Shop                   Boutique  
2   Japanese Restaurant       Thai Restaurant  Middle Eastern Restaurant  
3    Athletics & Sports           Coffee Shop                  Nightclub  
4                   Pub                  Café                      Hotel

In [152]:
neighborhoods_venues_sorted

Cluster Labels1  Cluster Labels  \
0                 0               0   
1                 3               3   
2                 4               4   
3                 2               2   
4                 0               0   
5                 0               0   
6                 0               0   
7                 0               0   
8                 0               0   
9                 0               0   
10                4               4   
11                4               4   
12                0               0   
13                1               1   
14                0               0   
15                0               0   
16                0               0   
17                0               0   
18                0               0   

                                         Neighborhood 1st Most Common Venue  \
0                                         Berczy Park           Coffee Shop   
1   CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
2                                  Central Bay Street           Coffee Shop   
3                                            Christie         Grocery Store   
4                                Church and Wellesley           Coffee Shop   
5                      Commerce Court, Victoria Hotel           Coffee Shop   
6              First Canadian Place, Underground city           Coffee Shop   
7                            Garden District, Ryerson        Clothing Store   
8   Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
9           Kensington Market, Chinatown, Grange Park                  Café   
10        Queen's Park, Ontario Provincial Government           Coffee Shop   
11                          Regent Park, Harbourfront           Coffee Shop   
12                           Richmond, Adelaide, King           Coffee Shop   
13                                           Rosedale                  Park   
14                                     St. James Town           Coffee Shop   
15                        St. James Town, Cabbagetown           Coffee Shop   
16                                     Stn A PO Boxes           Coffee Shop   
17           Toronto Dominion Centre, Design Exchange           Coffee Shop   
18                     University of Toronto, Harbord                  Café   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0           Cocktail Bar           Cheese Shop             Restaurant   
1         Airport Lounge      Airport Terminal                  Plane   
2         Sandwich Place    Italian Restaurant                   Café   
3                   Café                  Park            Candy Store   
4    Japanese Restaurant      Sushi Restaurant                Gay Bar   
5             Restaurant                 Hotel                   Café   
6                   Café                 Hotel    Japanese Restaurant   
7            Coffee Shop                  Café    Japanese Restaurant   
8               Aquarium                 Hotel                   Café   
9            Coffee Shop    Mexican Restaurant  Vietnamese Restaurant   
10      Sushi Restaurant           Yoga Studio         Discount Store   
11                Bakery                  Park                   Café   
12                  Café            Restaurant          Deli / Bodega   
13                 Trail            Playground           Dance Studio   
14                  Café             Gastropub           Cocktail Bar   
15            Restaurant                   Pub     Italian Restaurant   
16          Cocktail Bar    Seafood Restaurant                   Café   
17                 Hotel                  Café            Salad Place   
18             Bookstore   Japanese Restaurant     Italian Restaurant   

            5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Farmers Market    Seafood Restaurant              Beer Bar   
1                 H

In [149]:
neighborhoods_venues_sorted['1st Most Common Venue'].value_counts()

Coffee Shop        13
Café                2
Airport Service     1
Clothing Store      1
Grocery Store       1
Park                1
Name: 1st Most Common Venue, dtype: int64

In [150]:
neighborhoods_venues_sorted['2nd Most Common Venue'].value_counts()

Café                   4
Restaurant             2
Coffee Shop            2
Cocktail Bar           2
Hotel                  1
Sushi Restaurant       1
Sandwich Place         1
Bookstore              1
Aquarium               1
Trail                  1
Bakery                 1
Airport Lounge         1
Japanese Restaurant    1
Name: 2nd Most Common Venue, dtype: int64

In [151]:
neighborhoods_venues_sorted['3rd Most Common Venue'].value_counts()

Hotel                  3
Café                   2
Park                   2
Sushi Restaurant       1
Mexican Restaurant     1
Yoga Studio            1
Gastropub              1
Restaurant             1
Italian Restaurant     1
Japanese Restaurant    1
Pub                    1
Seafood Restaurant     1
Playground             1
Cheese Shop            1
Airport Terminal       1
Name: 3rd Most Common Venue, dtype: int64

### Coffee Shop is the most common venue in Neigborhood, which is not surprising considering that we analyzed Downtown Toronto.


## 4. Cluster Neighborhoods


Run _k_-means to cluster the neighborhood into 5 clusters.


In [120]:
# set number of clusters
kclusters = 5

Downtown_Toronto_grouped_clustering = Downtown_Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Downtown_Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 3, 4, 2, 0, 0, 0, 0, 0, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.


In [138]:

Downtown_Toronto_merged = Downtown_Toronto_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
Downtown_Toronto_merged= Downtown_Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')
Downtown_Toronto_merged # check the last columns!
Downtown_Toronto_merged.tail()

Postal Code           Borough                           Neighbourhood  \
14         M4W  Downtown Toronto                                Rosedale   
15         M5W  Downtown Toronto                          Stn A PO Boxes   
16         M4X  Downtown Toronto             St. James Town, Cabbagetown   
17         M5X  Downtown Toronto  First Canadian Place, Underground city   
18         M4Y  Downtown Toronto                    Church and Wellesley   

     Latitude  Longitude  Cluster Labels1  Cluster Labels  \
14  43.679563 -79.377529                1               1   
15  43.646435 -79.374846                0               0   
16  43.667967 -79.367675                0               0   
17  43.648429 -79.382280                0               0   
18  43.665860 -79.383160                0               0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
14                  Park                 Trail            Playground   
15           Coffee Shop          Cocktail Bar    Seafood Restaurant   
16           Coffee Shop            Restaurant                   Pub   
17           Coffee Shop                  Café                 Hotel   
18           Coffee Shop   Japanese Restaurant      Sushi Restaurant   

   4th Most Common Venue        5th Most Common Venue 6th Most Common Venue  \
14          Dance Studio  Eastern European Restaurant            Donut Shop   
15                  Café          Japanese Restaurant            Restaurant   
16    Italian Restaurant                       Bakery                  Café   
17   Japanese Restaurant                   Restaurant                   Gym   
18               Gay Bar                   Restaurant  Fast Food Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
14      Doner Restaurant               Dog Run   Distribution Center   
15                 Hotel    Italian Restaurant              Beer Bar   
16           Pizza Place              Pharmacy    Chinese Restaurant   
17   American Restaurant            Steakhouse         Deli / Bodega   
18           Yoga Studio                   Pub                  Café   

   10th Most Common Venue  
14         Discount Store  
15            Cheese Shop  
16                Butcher  
17       Asian Restaurant  
18                  Hotel

Finally, let's visualize the resulting clusters


In [140]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Downtown_Toronto_merged['Latitude'], Downtown_Toronto_merged['Longitude'], Downtown_Toronto_merged['Neighbourhood'], Downtown_Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### There are more frequent intervals within Clusters located closer to the harbour than other clusters

<a id='item5'></a>


## 5. Examine Clusters


#### Cluster 1


In [142]:
Downtown_Toronto_merged.loc[Downtown_Toronto_merged['Cluster Labels'] == 0, Downtown_Toronto_merged.columns[[1] + list(range(5, Downtown_Toronto_merged.shape[1]))]]

Borough  Cluster Labels1  Cluster Labels 1st Most Common Venue  \
2   Downtown Toronto                0               0        Clothing Store   
3   Downtown Toronto                0               0           Coffee Shop   
4   Downtown Toronto                0               0           Coffee Shop   
7   Downtown Toronto                0               0           Coffee Shop   
8   Downtown Toronto                0               0           Coffee Shop   
9   Downtown Toronto                0               0           Coffee Shop   
10  Downtown Toronto                0               0           Coffee Shop   
11  Downtown Toronto                0               0                  Café   
12  Downtown Toronto                0               0                  Café   
15  Downtown Toronto                0               0           Coffee Shop   
16  Downtown Toronto                0               0           Coffee Shop   
17  Downtown Toronto                0               0           Coffee Shop   
18  Downtown Toronto                0               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
2            Coffee Shop                  Café    Japanese Restaurant   
3                   Café             Gastropub           Cocktail Bar   
4           Cocktail Bar           Cheese Shop             Restaurant   
7                   Café            Restaurant          Deli / Bodega   
8               Aquarium                 Hotel                   Café   
9                  Hotel                  Café            Salad Place   
10            Restaurant                 Hotel                   Café   
11             Bookstore   Japanese Restaurant     Italian Restaurant   
12           Coffee Shop    Mexican Restaurant  Vietnamese Restaurant   
15          Cocktail Bar    Seafood Restaurant                   Café   
16            Restaurant                   Pub     Italian Restaurant   
17                  Café                 Hotel    Japanese Restaurant   
18   Japanese Restaurant      Sushi Restaurant                Gay Bar   

            5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
2       Middle Eastern Restaurant       Bubble Tea Shop                 Hotel   
3             American Restaurant    Seafood Restaurant        Cosmetics Shop   
4                  Farmers Market    Seafood Restaurant              Beer Bar   
7                  Clothing Store       Thai Restaurant                   Gym   
8                         Brewery            Restaurant   Fried Chicken Joint   
9                      Restaurant    Seafood Restaurant   Japanese Restaurant   
10                            Gym   American Restaurant    Seafood Restaurant   
11                            Bar                Bakery      Sushi Restaurant   
12  Vegetarian / Vegan Restaurant          Dessert Shop                   Bar   
15            Japanese Restaurant            Restaurant                 Hotel   
16                         Bakery                  Café           Pizza Place   
17                     Restaurant                   Gym   American Restaurant   
18                     Restaurant  Fast Food Restaurant           Yoga Studio   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
2         Cosmetics Shop        Lingerie Store     Italian Restaurant  
3             Restaurant    Italian Restaurant            Cheese Shop  
4                 Bakery        Breakfast Spot              Irish Pub  
7                  Hotel           Salad Place           Concert Hall  
8     Italian Restaurant        Scenic Lookout         History Museum  
9    American Restaurant    Italian Restaurant    Sporting Goods Shop  
10   Japanese Restaurant    Italian Restaurant          Deli / Bodega  
11        Sandwich Place                   Pub            College Gym  
12        Farmers Market                Bakery    Arts & Crafts Store  
15    Italian Restaurant    

Coffe Shops are the most common value following by restaurants, hotels, bars. The result is normal because downtown Toronto might be the most visited place both for tourists and locals.

#### Cluster 2


In [144]:
Downtown_Toronto_merged.loc[Downtown_Toronto_merged['Cluster Labels'] == 1, Downtown_Toronto_merged.columns[[1] + list(range(5, Downtown_Toronto_merged.shape[1]))]]

Borough  Cluster Labels1  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto                1               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14                 Trail            Playground          Dance Studio   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14  Eastern European Restaurant            Donut Shop      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14               Dog Run   Distribution Center         Discount Store

Cluster 2 has more social areas like parks, playgrounds and the transportation is good and can be done by rail. People might find this area as a good place to live in.

#### Cluster 3


In [143]:
Downtown_Toronto_merged.loc[Downtown_Toronto_merged['Cluster Labels'] == 2, Downtown_Toronto_merged.columns[[1] + list(range(5, Downtown_Toronto_merged.shape[1]))]]

Borough  Cluster Labels1  Cluster Labels 1st Most Common Venue  \
6  Downtown Toronto                2               2         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                  Café                  Park           Candy Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6            Restaurant    Italian Restaurant            Baby Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6    Athletics & Sports           Coffee Shop              Nightclub

#### Cluster 4


In [145]:
Downtown_Toronto_merged.loc[Downtown_Toronto_merged['Cluster Labels'] == 3, Downtown_Toronto_merged.columns[[1] + list(range(5, Downtown_Toronto_merged.shape[1]))]]

Borough  Cluster Labels1  Cluster Labels 1st Most Common Venue  \
13  Downtown Toronto                3               3       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13        Airport Lounge      Airport Terminal                 Plane   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13       Harbor / Marina      Sculpture Garden         Boat or Ferry   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13   Rental Car Location           Coffee Shop               Boutique

Cluster 4 is the airport area

#### Cluster 5


In [146]:
Downtown_Toronto_merged.loc[Downtown_Toronto_merged['Cluster Labels'] ==4, Downtown_Toronto_merged.columns[[1] + list(range(5, Downtown_Toronto_merged.shape[1]))]]

Borough  Cluster Labels1  Cluster Labels 1st Most Common Venue  \
0  Downtown Toronto                4               4           Coffee Shop   
1  Downtown Toronto                4               4           Coffee Shop   
5  Downtown Toronto                4               4           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                  Park                  Café   
1      Sushi Restaurant           Yoga Studio        Discount Store   
5        Sandwich Place    Italian Restaurant                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Pub        Breakfast Spot               Theater   
1                   Bar         Smoothie Shop    Italian Restaurant   
5       Bubble Tea Shop          Burger Joint           Salad Place   

  8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0             Wine Shop     Electronics Store      Performing Arts Venue  
1   Japanese Restaurant              Beer Bar             Sandwich Place  
5   Japanese Restaurant       Thai Restaurant  Middle Eastern Restaurant

Cluster 5 has coffe shops as the most commmon value behind the cluster 1 which has more coffe shops than cluster 5. Cluster 1 is less popular than Cluster 5. It might be the area where locals go outside and spend their lesiure times